python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -883,3 +929,145 @@ def _append_numpy_string_metadata(metadata: Dict[bytes, str], name: str, dtype:
         metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize // 4)
     elif dtype.type is np.bytes_:
         metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize)
+
+
+def _append_numpy_multidim_metadata(metadata: Dict[bytes, str], name: str, dtype: np.dtype) -> None:
+    """Append numpy multi-dimensional shapes to arrow metadata.
+
+    All column types are handled, but the metadata is only modified for
+    multi-dimensional columns.
+
+    Parameters
+    ----------
+    metadata : `dict` [`bytes`, `str`]
+        Metadata dictionary; modified in place.
+    name : `str`
+        Column name.
+    dtype : `np.dtype`
+        Numpy dtype.
+    """
+    if len(dtype.shape) > 1:
+        metadata[f"lsst::arrow::shape::{name}".encode("UTF-8")] = str(dtype.shape)
+
+
+def _multidim_shape_from_metadata(metadata: Dict[bytes, bytes], list_size: int, name: str) -> tuple[int, ...]:
+    """Retrieve the shape from the metadata, if available.
+
+    Parameters
+    ----------
+    metadata : `dict` [`bytes`, `bytes`]
+        Metadata dictionary.
+    list_size : `int`
+        Size of the list datatype.
+    name : `str`
+        Column name.
+
+    Returns
+    -------
+    shape : `tuple` [`int`]
+        Shape associated with the column.
+
+    Raises
+    ------
+    RuntimeError
+        Raised if metadata is found but has incorrect format.
+    """
+    md_name = f"lsst::arrow::shape::{name}"
+    if (encoded := md_name.encode("UTF-8")) in metadata:
+        groups = re.search(r"\((.*)\)", metadata[encoded].decode("UTF-8"))
+        if groups is None:
+            raise RuntimeError("Illegal value found in metadata.")
+        shape = tuple(int(x) for x in groups[1].split(",") if x != "")
```

Comment:
Using a regex instead of `startswith("(")` and `endswith(")")` but not instead of `split(",")` probably puts you in a sparsely-populated region of regex love-hate space, but I won't argue with it (especially since I'm not sure how much you have to worry about whitespace or something outside those parens).  If you'd like a regex for the whole thing I'm sure I could cook one up.

Reply:
I think I'll leave it as is, but interestingly this is essentially the same solution that chatgpt recommended when I just asked.  So it must be common online?

Reply:
To be clear, I don't use chatgpt for programming advice, I was just curious what it would say.